### PR TITLE
KBFS archive frontend first pass

### PR DESCRIPTION
**Pull request by songgao**
@chrisnojima @strib

### CODE REVIEW

File: go/kbfs/simplefs/archive.go [hunk 1]
```diff
@@ -829,7 +858,18 @@ func (m *archiveManager) doZipping(ctx context.Context, jobID string) (err error
 
 		err = zipWriterAddDir(ctx, zipWriter, workspaceDir, updateBytesZipped)
 		if err != nil {
-			return fmt.Errorf("zipWriter.AddFS to %s error: %v", jobDesc.ZipFilePath, err)
+			return fmt.Errorf("zipWriterAddDir into %s error: %v", jobDesc.ZipFilePath, err)
+		}
+
+		{ // write the manifest
```
Comment (songgao) [inline go/kbfs/simplefs/archive.go:864]: I forgot this in the backend MR

File: go/kbfs/simplefs/archive.go [hunk 2]
```diff
@@ -829,7 +858,18 @@ func (m *archiveManager) doZipping(ctx context.Context, jobID string) (err error
...
+			w, err := zipWriter.Create("manifest.json")
```
Comment (strib) [inline go/kbfs/simplefs/archive.go:865]: Not sure how this works, but do you need to close `w`?

File: go/kbfs/simplefs/simplefs.go [hunk 3]
```diff
@@ -3700,33 +3698,76 @@ func (k *SimpleFS) SimpleFSGetArchiveStatus(ctx context.Context) (
...
+// SimpleFSGetArchiveStatus implements the SimpleFSInterface.
+func (k *SimpleFS) SimpleFSGetArchiveStatus(ctx context.Context) (
+	status keybase1.SimpleFSArchiveStatus, err error) {
+	ctx = k.makeContext(ctx)
+	state, errorStates := k.archiveManager.getCurrentState(ctx)
+	return k.archiveStateToStatus(ctx, state, errorStates)
+}
+
+// SimpleFSGetArchiveJobFreshness implements the SimpleFSInterface.
+func (k *SimpleFS) SimpleFSGetArchiveJobFreshness(ctx context.Context, jobID string) (keybase1.SimpleFSArchiveJobFreshness, error) {
```
Comment (strib) [inline go/kbfs/simplefs/simplefs.go:3724]: Is it intentional that the command line implementation is still only calling the old endpoint, and so is no longer getting freshness information like it was before?  It looks like it depends on the `CurrentTLFRevision` field that is no longer being filled in by the other call.

File: go/libkb/notify_router.go [hunk 4]
```diff
@@ -939,6 +942,28 @@ func (n *NotifyRouter) HandleFSSubscriptionNotifyPath(arg keybase1.FSSubscriptio
...
+func (n *NotifyRouter) HandleSimpleFSArchiveStatusChanged(ctx context.Context, status keybase1.SimpleFSArchiveStatus) {
+	if n == nil {
+		return
+	}
+	// For all connections we currently have open...
+	n.cm.ApplyAll(func(id ConnectionID, xp rpc.Transporter) bool {
+		// If the connection wants the `Kbfs` notification type
+		if n.getNotificationChannels(id).Notifysimplefs { // In the background do...
+			go func() {
+				// A send of a `FSSyncActivity` RPC with the notification
```
Comment (strib) [inline go/libkb/notify_router.go:954]: Copied comment.

**Review by strib — COMMENTED** [review]
A few nits/questions, the Go side looks fine overall though.

File: go/kbfs/simplefs/archive.go (hunk 2 above)
Comment (songgao) [reply go/kbfs/simplefs/archive.go:865]: Yeah this is a bit odd. This gives an io.Writer not an io.WriteCloser. I think the zip writer close is what flushes.

File: go/kbfs/simplefs/simplefs.go (hunk 3 above)
Comment (songgao) [reply go/kbfs/simplefs/simplefs.go:3724]: Good catch! I forgot; fixed

File: shared/common-adapters/progress-bar.tsx [hunk 5]
```diff
@@ -5,16 +5,27 @@ type Props = {
...
+    ...(flatRight ? styles.flatRight : {}),
     width: `${Math.max(0, Math.min(1, ratio)) * 100}%`,
   }
   return (
-    <Box style={Styles.collapseStyles([styles.outer, style])}>
+    <Box
+      style={Styles.collapseStyles([
+        styles.outer,
+        style,
+        flatLeft ? styles.flatLeft : {},
```
Comment (chrisnojima) [inline shared/common-adapters/progress-bar.tsx:25]: `flatLeft && styles.flatLeft`. collapseStyles expects false and will maybe deopt you if you pass in new objects each time

File: shared/constants/archive.tsx [hunk 6]
```diff
@@ -47,6 +73,44 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
 
+  const setKBFSJobStatus = (status: T.RPCGen.SimpleFSArchiveStatus) =>
+    set(s => ({
+      ...s,
+      kbfsJobs: new Map(
+        // order is retained
+        (status.jobs || []).map(job => [
+          job.desc.jobID,
+          {
+            id: job.desc.jobID,
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:84]: lint will complain the keys aren't ordered here

File: shared/constants/archive.tsx [hunk 7]
```diff
@@ -84,6 +146,27 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
+    },
+    loadKBFS: async () => {
+      const status = await T.RPCGen.SimpleFSSimpleFSGetArchiveStatusRpcPromise()
+      setKBFSJobStatus(status)
+    },
+    loadKBFSJobFreshness: async (jobID: string) => {
+      const resp = await T.RPCGen.SimpleFSSimpleFSGetArchiveJobFreshnessRpcPromise({jobID})
+      set(s => ({
+        ...s,
+        kbfsJobsFreshness: new Map([...s.kbfsJobsFreshness, [jobID, resp.currentTLFRevision]]), // ordering doesn't matter here
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:160]: with immer you can just set it
```
s. kbfsJobsFreshness.set(jobID, resp.currentTLFRevision)
```
else you'll get a new map each time possibly

File: shared/constants/archive.tsx (hunk 6 above)
Comment (chrisnojima) [reply shared/constants/archive.tsx:84]: you can just call set here (see below)

File: shared/constants/notifications.tsx [hunk 8]
```diff
@@ -126,6 +126,7 @@ export const _useState = Z.createZustand<State>((set, get) => {
               kbfslegacy: false,
               kbfsrequest: false,
               kbfssubscription: true,
+              notifysimplefs: true,
```
Comment (chrisnojima) [inline shared/constants/notifications.tsx:129]: key out of order

File: shared/settings/archive/index.tsx [hunk 9]
```diff
@@ -84,13 +84,142 @@ const Job = React.memo(function Job(p: {index: number; id: string}) {
   )
 })
 
+const KBFSJob = React.memo(function Job(p: {index: number; id: string}) {
```
Comment (chrisnojima) [inline shared/settings/archive/index.tsx:87]: should call this function `KBFSJob` also

File: shared/settings/archive/index.tsx [hunk 10]
```diff
@@ -102,7 +231,9 @@ const Archive = C.featureFlags.archive
       const clearCompleted = C.useArchiveState(s => s.dispatch.clearCompleted)
 
       const jobMap = C.useArchiveState(s => s.jobs)
+      const kbfsJobMap = C.useArchiveState(s => s.kbfsJobs)
       const jobs = React.useMemo(() => [...jobMap.keys()], [jobMap])
+      const kbfsJobs = React.useMemo(() => [...kbfsJobMap.keys()], [jobMap])
```
Comment (chrisnojima) [inline shared/settings/archive/index.tsx:236]: i don't think you really need this to be useMemo, its not a dependency of anything, if kbfsJobMap changes then this does to so you're not skipping any rendering

File: shared/common-adapters/progress-bar.tsx [hunk 11]
```diff
@@ -37,6 +48,8 @@ const styles = Styles.styleSheetCreate(() => ({
       boxShadow: `inset 0 1px 0 0 ${Styles.globalColors.black_05}`,
     },
   }),
+  flatLeft: {borderBottomLeftRadius: 0, borderTopLeftRadius: 0},
```
Comment (chrisnojima) [inline shared/common-adapters/progress-bar.tsx:51]: sort keys

File: shared/constants/archive.tsx [hunk 12]
```diff
@@ -47,6 +73,44 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
+              [T.RPCGen.SimpleFSArchiveJobPhase.indexing]: 'Indexing',
+              [T.RPCGen.SimpleFSArchiveJobPhase.indexed]: 'Indexed',
+              [T.RPCGen.SimpleFSArchiveJobPhase.copying]: 'Copying',
+              [T.RPCGen.SimpleFSArchiveJobPhase.copied]: 'Copied',
+              [T.RPCGen.SimpleFSArchiveJobPhase.zipping]: 'Zipping',
+              [T.RPCGen.SimpleFSArchiveJobPhase.done]: 'Done',
+            }[job.phase],
+            kbfsPath: job.desc.kbfsPathWithRevision.path,
+            kbfsRevision:
+              job.desc.kbfsPathWithRevision.archivedParam.KBFSArchivedType ==
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:97]: `===`

File: shared/constants/archive.tsx [hunk 13]
```diff
@@ -84,6 +146,27 @@ export const _useState = Z.createZustand<State>((set, get) => {
         }
       })
     },
+    cancelOrDismissKBFS: async (jobID: string) => {
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:149]: i'd make this not async and make an async helper inside that you `ignorePromise` on, currently this return Promise<void> vs void due to this

File: shared/constants/archive.tsx [hunk 14]
```diff
@@ -84,6 +146,27 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
+      await T.RPCGen.SimpleFSSimpleFSArchiveCancelOrDismissJobRpcPromise({jobID})
+    },
+    loadKBFS: async () => {
+      const status = await T.RPCGen.SimpleFSSimpleFSGetArchiveStatusRpcPromise()
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:153]: same as above

File: shared/constants/archive.tsx [hunk 15]
```diff
@@ -84,6 +146,27 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
+      setKBFSJobStatus(status)
+    },
+    loadKBFSJobFreshness: async (jobID: string) => {
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:156]: same here

File: shared/constants/archive.tsx [hunk 16]
```diff
@@ -84,6 +146,27 @@ export const _useState = Z.createZustand<State>((set, get) => {
...
+      set(s => ({
+        ...s,
+        kbfsJobsFreshness: new Map([...s.kbfsJobsFreshness, [jobID, resp.currentTLFRevision]]), // ordering doesn't matter here
+      }))
+    },
+    onEngineIncoming: action => {
+      switch (action.type) {
+        case EngineGen.keybase1NotifySimpleFSSimpleFSArchiveStatusChanged:
+          setKBFSJobStatus(action.payload.params.status)
+          break
```
Comment (chrisnojima) [inline shared/constants/archive.tsx:167]: add default

**Comment by chrisnojima** [comment]
run `yarn lint` for some issues

**Review by chrisnojima — APPROVED** [review]

**Pull request merged**